### ISSUE TITLE
Error compiling magnus v0.5.1 on arm-linux

### ISSUE THREAD

**Issue by yammine** [matsadler/magnus#63]
Hello 👋 
First off, thank you for this project it's great.

I've noticed a compilation error on `arm-linux` after updating from version `0.4` to `0.5` 

The error:
```rust
error[E0308]: mismatched types
   --> /home/runner/work/roaring-rb/roaring-rb/tmp/cargo-vendor/magnus/src/r_bignum.rs:295:13
    |
291 |     fn sign(self) -> u64 {
    |                      --- expected `u64` because of return type
...
295 |             r_basic.as_ref().flags & (ruby_fl_type::RUBY_FL_USER1 as VALUE)
    |             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expected `u64`, found `u32`
    |
help: you can convert a `u32` to a `u64`
    |
295 |             (r_basic.as_ref().flags & (ruby_fl_type::RUBY_FL_USER1 as VALUE)).into()
    |             +                                                               ++++++++
```
Here is a link to the GH Action run where I've observed this: [link ](https://github.com/yammine/roaring-rb/actions/runs/4300924191/jobs/7497598198)
All other platforms successfully compile `magnus` so I think it's isolated to the `arm-linux` platform.

If someone knows off the top of their head how to fix this I'd love to meaningfully contribute by submitting a patch.

Thanks for your attention

**Comment by matsadler** [comment]
Hey, thanks for the bug report.

So the issue here is `arm-linux` is a 32 bit platform (`aarch64-linux` is 64 bit arm), and when I wrote this code I wasn't thinking about 32 bit platforms (I'm happy to get Magnus compiling for 32 bit platforms, but no guarantee it'll actually do what it's meant to do).

I'll explain what's going on from the ground up so that you have the context if you'd like to put together a patch. Sorry if any of this is stuff you know already.

In Ruby's internals / the Ruby C API all Ruby objects are represented by `VALUE`, this is effectively a pointer to the object (all of Magnus's types are just wrappers around a `VALUE`).

Ruby defines `VALUE` as an integer the size of a pointer, which is basically Rust's `usize`. But something gets lost in translation so it ends up in Rust as `VALUE` being defined as an alias to `u64` on 64 bit platforms, and an alias to `u32` on 32 bit platforms.

So `VALUE` is a pointer to a Ruby object, and `r_basic` is the thing it points to. `r_basic` usually holds some very basic information about the object and then further pointers to specifics. One of the basic bits of data is `flags` this is a [bitfield](https://en.wikipedia.org/wiki/Bit_field) that encodes some information about the object. In the case of `RBignum` one of the things in the `flags` is if the bignum is positive or negative. For some weird historical reason Ruby defines the type of `flags` as `VALUE`, even though it's not a Ruby object (relying on knowing that under the hood `VALUE` is just an integer).

To check the bit in `flags` that represents whether the bignum is positive or negative we `&` (bitwise and) the flags with the constant `ruby_fl_type::RUBY_FL_USER1`. The type of this *isn't* `VALUE`, but Rust's `&` operator wants the types on both sides to be the same, so we cast it to `VALUE`. As both sides of the `&` are a `VALUE`, the end result is `VALUE`. As the result is most definitely not a Ruby object I didn't want to use `VALUE` as the return type, so used `u64`.

This works because on all the platforms I test on `VALUE` is an alias of `u64`, but it breaks on `arm-linux` because there `VALUE` is an alias of `u32`, which then won't match the return type.

One other detail, `sign`/`is_positive`/`is_negative` exist in Ruby's C API as macros, which can't be used from Rust, so have to be reimplemented for Rust. The `sign` function isn't public, it's only here to implement `is_positive`. While `sign` does exist in Ruby's C API I didn't have any intention of exposing it publicly as it's just kind of weird, like it's called `sign` and the return value is `0` for negative or any other number for positive, how does that make sense? So `sign` doesn't need to exist, can have a different return type, whatever, along as `is_positive` stays working the same.


I can think of a few different ways to fix this, I don't have any strong reason to prefer one over another, and you might be able to think of another. Hopefully that's enough to go off without just telling you exactly what to do (which seems like it'd be boring for you).

**Comment by yammine** [comment]
Thank you for the very thoughtful reply! :) 
I didn't know any of this, so it was very educational for me. 

I actually would appreciate some pointers on where I might begin to implement support for 32bit platforms, though no promises on my ability to actually get it done lol

I'll just drop support for that platform for my gem in the meantime

**Comment by matsadler** [comment]
Hey, sorry if you've not had time to contribute. I went ahead and fixed this in 7f42447 and put out a 0.5.2 release. I went with the 'inline `sign` to `is_positive` approach.
